File: playground/src/main/java/com/jiamian/translation/service/ModelApiService.java
package com.jiamian.translation.service;

import cn.hutool.core.collection.CollectionUtil;
import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.google.common.collect.Lists;
import com.jiamian.translation.dao.repository.MetaRepository;
import com.jiamian.translation.dao.repository.ModelCreatorRepository;
import com.jiamian.translation.dao.repository.ModelRepository;
import com.jiamian.translation.dao.repository.ModelTagsRepository;
import com.jiamian.translation.entity.dto.CreatorApiDTO;
import com.jiamian.translation.entity.dto.FilesApiDTO;
import com.jiamian.translation.entity.dto.ImagesApiDTO;
import com.jiamian.translation.entity.dto.MetaApiDTO;
import com.jiamian.translation.entity.dto.ModelApiDTO;
import com.jiamian.translation.entity.dto.ModelVersionsApiDTO;
import com.jiamian.translation.entity.dto.StatusApiDTO;
import com.jiamian.translation.model.Meta;
import com.jiamian.translation.model.Model;
import com.jiamian.translation.model.ModelCreator;
import com.jiamian.translation.model.ModelTags;
import com.jiamian.translation.util.DateUtil;
import com.jiamian.translation.util.QiniuUtil;
import org.apache.commons.lang3.ObjectUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.io.ByteArrayInputStream;
import java.util.Arrays;
import java.util.List;

/**
 * @author DingGuangHui
 * @date 2023/3/28
 */
@Service
public class ModelApiService {
	@Autowired
	private ModelRepository modelRepository;
	@Autowired
	private MetaRepository metaRepository;
	@Autowired
	private ModelCreatorRepository modelCreatorRepository;
	@Autowired
	private ModelTagsRepository modelTagsRepository;

	@Autowired
	private QiNiuService qiNiuService;

	private static final String MODELS_JSON_FILE_NAME = "%s_models.json";
	private static final String MODEL_DETAIL_URL = "https://models.paomiantv.cn/models/Detail?id=%s";

	/**
	 * 当前model信息构造json上传七牛云
	 */
	public void uploadQiniuModelsJson() {
		List<ModelApiDTO> models = Lists.newArrayList();
		String fileName = String.format(MODELS_JSON_FILE_NAME, DateUtil.formatYYYYMMDD());

		Model model = new Model();
		model.setStatus(1);
		int pageSize = 10;
		int pageIdx = 0;

		List<Model> dbModels;
		List<ModelApiDTO> item = Lists.newArrayList();

		do {
			Page<Model> p = modelRepository.findAll(Example.of(model), PageRequest.of(pageIdx, pageSize, Sort.Direction.ASC,
					"id"));
			dbModels = p.getContent();
			pageIdx++;


			for (Model dbModel : dbModels) {
				ModelApiDTO modelApiDTO = new ModelApiDTO();
				modelApiDTO.setId(dbModel.getId());
				modelApiDTO.setName(dbModel.getName());
				modelApiDTO.setDescription(dbModel.getDescription());
				modelApiDTO.setType(dbModel.getType());
				String trainedWords = "";
				String baseModel = "";

				String downloadUrl = String.format(MODEL_DETAIL_URL, dbModel.getModelId());

				StatusApiDTO statusApiDTO = new StatusApiDTO();
				statusApiDTO.setDownloadCount(dbModel.getDownloadCount());
				statusApiDTO.setFavoriteCount(dbModel.getFavoriteCount());
				statusApiDTO.setCommentCount(dbModel.getCommentCount());
				statusApiDTO.setRatingCount(dbModel.getRatingCount());
				statusApiDTO.setRating(Double.parseDouble(dbModel.getRating()));
				modelApiDTO.setStats(statusApiDTO);

				CreatorApiDTO creatorApiDTO = new CreatorApiDTO();
				List<ModelCreator> creators = modelCreatorRepository.findByModelId(dbModel.getModelId());
				if (CollectionUtil.isNotEmpty(creators)) {
					creatorApiDTO.setUsername(creators.get(0).getUsername());
					creatorApiDTO.setImage(creators.get(0).getImage());
				}
				modelApiDTO.setCreator(creatorApiDTO);


				List<ModelTags> modelTags = modelTagsRepository.findByModelId(dbModel.getModelId());
				if (CollectionUtil.isNotEmpty(modelTags)) {
					ModelTags modelT = modelTags.get(0);

					String tags = modelT.getTagText();
					trainedWords = ObjectUtils.isNotEmpty(modelT.getTrainedWords()) ? modelT.getTrainedWords() : "";
					baseModel = modelT.getBaseModel();
					modelApiDTO.setTags(Arrays.asList(tags.split(",")));
				}


				ModelVersionsApiDTO mvDTO = new ModelVersionsApiDTO();
				mvDTO.setDownloadUrl(downloadUrl);
				mvDTO.setId(dbModel.getId());
				mvDTO.setModelId(dbModel.getModelId());
				mvDTO.setName(dbModel.getName());
				mvDTO.setCreateAt(dbModel.getCreateDate());
				mvDTO.setUpdatedAt(dbModel.getCreateDate());
				mvDTO.setTrainedWords(Lists.newArrayList(trainedWords.split(",")));
				mvDTO.setBaseModel(baseModel);

				FilesApiDTO filesApiDTO = new FilesApiDTO();
				filesApiDTO.setDownloadUrl(downloadUrl);
				mvDTO.setFiles(Lists.newArrayList(filesApiDTO));


				List<Meta> metas = metaRepository.findByModelId(dbModel.getModelId());
				List<ImagesApiDTO> images = Lists.newArrayList();

				if (CollectionUtil.isNotEmpty(metas)) {
					for (Meta meta : metas) {
						ImagesApiDTO imagesApiDTO = new ImagesApiDTO();
						imagesApiDTO.setUrl(meta.getQiniuUrl());
						imagesApiDTO.setNsfw(false);

						JSONObject imgInfo = QiniuUtil.getImgInfo(meta.getQiniuUrl());
						Integer width = imgInfo.getInteger("width");
						Integer height = imgInfo.getInteger("height");
						imagesApiDTO.setWidth(width);
						imagesApiDTO.setHeight(height);

						MetaApiDTO metaApiDTO = new MetaApiDTO();

						metaApiDTO.setSize(width + "x" + height);
						metaApiDTO.setSeed(Long.parseLong(meta.getSeed()));
						metaApiDTO.setModel(dbModel.getName());
						metaApiDTO.setSteps(Integer.parseInt(meta.getSteps()));
						metaApiDTO.setPrompt(meta.getPrompt());
						metaApiDTO.setSampler(meta.getSampler());
						metaApiDTO.setCfgScale(Double.parseDouble(meta.getCfgScale()));
						metaApiDTO.setNegativePrompt(meta.getNegativePrompt());

						imagesApiDTO.setMeta(metaApiDTO);
						images.add(imagesApiDTO);
					}
				}

				mvDTO.setImages(images);
				modelApiDTO.setModelVersions(Lists.newArrayList(mvDTO));
				item.add(modelApiDTO);
			}
			break;
		} while (CollectionUtil.isNotEmpty(dbModels));

		ApiResp apiResp = new ApiResp(item);

		qiNiuService.uploadFile(new ByteArrayInputStream(JSON.toJSONString(apiResp).getBytes()),
				fileName);
	}

	class ApiResp {
		List<ModelApiDTO> items = Lists.newArrayList();

		public ApiResp() {
		}

		public ApiResp(List<ModelApiDTO> items) {
			this.items = items;
		}

		public List<ModelApiDTO> getItems() {
			return items;
		}

		public void setItems(List<ModelApiDTO> items) {
			this.items = items;
		}
	}
}

File: playground/src/main/java/com/jiamian/translation/service/ModelApiService.kt
package com.jiamian.translation.service

import cn.hutool.core.collection.CollectionUtil
import com.google.common.collect.Lists
import com.jiamian.translation.dao.repository.MetaRepository
import com.jiamian.translation.dao.repository.ModelCreatorRepository
import com.jiamian.translation.dao.repository.ModelRepository
import com.jiamian.translation.dao.repository.ModelTagsRepository
import com.jiamian.translation.entity.dto.api.*
import com.jiamian.translation.model.Model
import com.jiamian.translation.util.QiniuUtil.getImgInfo
import org.apache.commons.lang3.ObjectUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Example
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import java.util.*

/**
 * @author DingGuangHui
 * @date 2023/3/28
 */
@Service
class ModelApiService {
    @Autowired
    private lateinit var modelRepository: ModelRepository

    @Autowired
    private lateinit var metaRepository: MetaRepository

    @Autowired
    private lateinit var modelCreatorRepository: ModelCreatorRepository

    @Autowired
    private lateinit var modelTagsRepository: ModelTagsRepository


    /**
     * 当前model信息构造json上传七牛云
     */
    fun pageModelApi(pageSize: Int, pageIdx: Int): ApiResp {
        val example = Model()
        example.status = 1

        val item: MutableList<ModelApiDTO> = Lists.newArrayList()

        val p = modelRepository.findAll(Example.of(example),
                PageRequest.of(pageIdx, pageSize, Sort.Direction.ASC,
                        "id"))

        for (dbModel in p.content) {
            val modelApiDTO = ModelApiDTO(dbModel.id)
            modelApiDTO.name = dbModel.name
            modelApiDTO.description = dbModel.description
            modelApiDTO.type = dbModel.type
            var trainedWords = ""
            var baseModel = ""
            val downloadUrl = String.format(MODEL_DETAIL_URL,
                    dbModel.modelId)


            val statusApiDTO = StatusApiDTO()
            statusApiDTO.downloadCount = dbModel.downloadCount
            statusApiDTO.favoriteCount = dbModel.favoriteCount
            statusApiDTO.commentCount = dbModel.commentCount
            statusApiDTO.ratingCount = dbModel.ratingCount
            statusApiDTO.rating = dbModel.rating.toDouble()
            modelApiDTO.stats = statusApiDTO


            val creatorApiDTO = CreatorApiDTO()
            val creators = modelCreatorRepository.findByModelId(dbModel.modelId)
            if (CollectionUtil.isNotEmpty(creators)) {
                creatorApiDTO.username = creators[0]!!.username
                creatorApiDTO.image = creators[0]!!.image
            }
            modelApiDTO.creator = creatorApiDTO


            val modelTags = modelTagsRepository
                    .findByModelId(dbModel.modelId)
            if (CollectionUtil.isNotEmpty(modelTags)) {
                val modelT = modelTags[0]
                val tags = modelT!!.tagText
                trainedWords = if (ObjectUtils.isNotEmpty(
                                modelT.trainedWords)) modelT.trainedWords else ""
                baseModel = modelT.baseModel
                modelApiDTO.tags = Arrays.asList(*tags.split(",".toRegex()).toTypedArray())
            }


            val mvDTO = ModelVersionsApiDTO()
            mvDTO.downloadUrl = downloadUrl
            mvDTO.id = dbModel.id
            mvDTO.modelId = dbModel.modelId
            mvDTO.name = dbModel.name
            mvDTO.createAt = dbModel.createDate
            mvDTO.updatedAt = dbModel.createDate
            mvDTO.trainedWords = Lists.newArrayList(*trainedWords.split(",".toRegex()).toTypedArray())
            mvDTO.baseModel = baseModel
            val filesApiDTO = FilesApiDTO()
            filesApiDTO.downloadUrl = downloadUrl
            mvDTO.files = Lists.newArrayList(filesApiDTO)
            val metas = metaRepository
                    .findByModelId(dbModel.modelId)
            val images: MutableList<ImagesApiDTO> = Lists.newArrayList()
            if (CollectionUtil.isNotEmpty(metas)) {
                for (meta in metas) {
                    val imagesApiDTO = ImagesApiDTO()
                    imagesApiDTO.url = meta!!.qiniuUrl + "-mshalf"
                    imagesApiDTO.nsfw = false
                    val imgInfo = getImgInfo(meta.qiniuUrl)
                    val width = imgInfo.getInteger("width")
                    val height = imgInfo.getInteger("height")
                    imagesApiDTO.width = width
                    imagesApiDTO.height = height
                    val metaApiDTO = MetaApiDTO()
                    metaApiDTO.Size = width.toString() + "x" + height
                    metaApiDTO.seed = meta.seed.toLong()
                    metaApiDTO.Model = dbModel.name
                    metaApiDTO.steps = meta.steps.toInt()
                    metaApiDTO.prompt = meta.prompt
                    metaApiDTO.sampler = meta.sampler
                    metaApiDTO.cfgScale = meta.cfgScale.toDouble()
                    metaApiDTO.negativePrompt = meta.negativePrompt
                    imagesApiDTO.meta = metaApiDTO
                    images.add(imagesApiDTO)
                }
            }
            mvDTO.images = images
            modelApiDTO.modelVersions = Lists.newArrayList(mvDTO)
            item.add(modelApiDTO)
        }

        return ApiResp(item, MetaDataApiDTO(p.totalElements, pageIdx, pageSize, p.totalPages))
    }

    inner class ApiResp {
        var items: List<ModelApiDTO> = Lists.newArrayList()
        var metadata: MetaDataApiDTO = MetaDataApiDTO(0, 0, 0, 0)

        constructor() {}
        constructor(items: List<ModelApiDTO>, metadata: MetaDataApiDTO) {
            this.items = items
            this.metadata = metadata
        }
    }

    companion object {
        //
        private const val MODEL_DETAIL_URL = "https://models.paomiantv.cn/models/Detail?id=%s"
    }
}